File: Batch1/Trees/Binary Tree/Questions/DiameterQuestions.java
class DiameterQuestions {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    // Diameter of a binary Tree (Leetcode 543)===================
    public int heightOfBinaryTree(TreeNode root){
        if(root == null){
            return -1;
        }

        int leftHeight = heightOfBinaryTree(root.left);
        int rightHeight = heightOfBinaryTree(root.right);

        return Math.max(leftHeight, rightHeight) + 1;
    }

    // O(N^2)
    public int diameterOfBinaryTree(TreeNode root) {
        if(root == null) return -1;

        int leftHeight = heightOfBinaryTree(root.left);
        int rightHeight = heightOfBinaryTree(root.right);

        int currentNodeDiameter = leftHeight + rightHeight + 2;

        int leftMaxDiameter = diameterOfBinaryTree(root.left);
        int righttMaxDiameter = diameterOfBinaryTree(root.right);

        return Math.max(Math.max(leftMaxDiameter, righttMaxDiameter), currentNodeDiameter);
    }

    // Method 2 , {height, diameter}
    public int[] heightDiameterOfBinaryTree(TreeNode root){
        if(root == null){
            return new int[]{-1,-1};
        }

        int[] leftHeightDia = heightDiameterOfBinaryTree(root.left);
        int[] rightHeightDia = heightDiameterOfBinaryTree(root.right);

        int leftHeight = leftHeightDia[0];
        int rightHeight = rightHeightDia[0];

        int leftMaxDiameter = leftHeightDia[1];
        int rightMaxDiameter = rightHeightDia[1];

        int currentNodeHeight = Math.max(leftHeight, rightHeight) + 1;
        int currentNodeDiameter = leftHeight + rightHeight + 2;

        int maxDiameter = Math.max(Math.max(leftMaxDiameter, rightMaxDiameter), currentNodeDiameter);

        return new int[]{currentNodeHeight, maxDiameter};
    }

    public int diameterOfBinaryTree(TreeNode root) {
        return heightDiameterOfBinaryTree(root)[1];
    }

    // Method 3 ==================================================
    // int maxDia;
    public int heightOfBinaryTree(TreeNode root, int[] maxDia){
        if(root == null){
            return -1;
        }

        int leftHeight = heightOfBinaryTree(root.left, maxDia);
        int rightHeight = heightOfBinaryTree(root.right, maxDia);

        int currentNodeDiameter = leftHeight + rightHeight + 2;
        maxDia[0] = Math.max(maxDia[0], currentNodeDiameter);
        // maxDia = Math.max(maxDia, currentNodeDiameter);

        return Math.max(leftHeight, rightHeight) + 1;
    }

    public int diameterOfBinaryTree(TreeNode root) {
        // maxDia = -1;
        int[] maxDia = new int[0];
        heightOfBinaryTree(root, maxDia);

        return maxDia[0];
    }

    // Max path sum from leaf to leaf (https://www.geeksforgeeks.org/problems/maximum-path-sum/1) ===========================
    class Solution {
        int maxLeafToNode(Node root, int[] maxLeafToLeaf){
            if(root == null){
                return -(int)(1e8);
            }

            if(root.left == null && root.right == null){
                return root.data;
            }

            int maxLeftLTN = maxLeafToNode(root.left, maxLeafToLeaf);
            int maxRightLTN = maxLeafToNode(root.right, maxLeafToLeaf);

            if(root.left != null && root.right != null){
                maxLeafToLeaf[0] = Math.max(maxLeafToLeaf[0], maxLeftLTN + maxRightLTN + root.data);
            }

            return Math.max(maxLeftLTN, maxRightLTN) + root.data;
        }


        int maxPathSum(Node root) {
            int[] maxLeafToLeaf = new int[]{-(int)(1e8)};

            int leafToRoot = maxLeafToNode(root, maxLeafToLeaf);
            if(maxLeafToLeaf[0] == -(int)(1e8)){ // to handle one TC of Tree with only right/left nodes
                return leafToRoot;
            }

            return maxLeafToLeaf[0];
        }
    }























    public static void main(String[] args){

    }
}

File: Batch1/Trees/Binary Tree/Questions/diameterQuestions.cpp
#include<vector>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};


// Diameter of a binary Tree (Leetcode 543)===================

// Method 3, for first 2 methods see Java

int heightOfBinaryTree(TreeNode* root, int& maxDia){
    if(root == NULL){
        return -1;
    }

    int leftHeight = heightOfBinaryTree(root->left, maxDia);
    int rightHeight = heightOfBinaryTree(root->right, maxDia);

    int currentNodeDiameter = leftHeight + rightHeight + 2;
    maxDia = max(maxDia, currentNodeDiameter);

    return max(leftHeight, rightHeight) + 1;
}

int diameterOfBinaryTree(TreeNode* root) {
    int maxDia = -1;
    heightOfBinaryTree(root, maxDia);
    return maxDia;
}


// Leetcode 124 (Node to Node max path sum) ==
int maxNodeToNode(TreeNode* root, int& ans){
    if(root == nullptr){
        return -(int)(1e8);
    }

    int maxLeftPathSum = maxNodeToNode(root->left, ans);
    int maxRightPathSum = maxNodeToNode(root->right, ans);


    int maxPathSum = max(max(maxLeftPathSum + root->val, maxRightPathSum + root->val), root->val);
    // 3 Scenarios -> left Path + current node, right path + current node, current node

    int maxNodeToNode = max(maxPathSum, maxLeftPathSum + maxRightPathSum + root->val);
    ans = max(ans, maxNodeToNode);
    // all the above 3 scenariors, leftPathSum + rightPathSum + root.val

    return maxPathSum;
}

int maxPathSum(TreeNode* root) {
    int ans = -(int)(1e8);
    maxNodeToNode(root,ans);
    return ans;
}


















int main(){
    return 0;
}